mono/metadata/threads.c

```diff
@@ -5984,29 +6014,120 @@ mono_threads_summarize (MonoContext *ctx, gchar **out, MonoStackHash *hashes)
 			size_t old_num_summarized = num_threads_summarized;
 
 			sigprocmask (SIG_UNBLOCK, &sigset, &old_sigset);
+
+			MonoBoolean success;
+			ves_icall_System_Threading_Interlocked_CompareExchange_Int_Success (&summarizing_thread_state, MONO_SUMMARY_EXPECT /* set */, MONO_SUMMARY_EMPTY /* compare */, &success);
+			if (!success)
+				g_error ("Summary memory was not in a clean state prior to entry");
+
+			if (summarizing_thread_state != MONO_SUMMARY_EXPECT) {
+				const char *name = thread_summary_state_to_str (summarizing_thread_state);
+				g_error ("Status after init wrong: %s\n", name);
+			}
+
+			summarizing_thread = tid;
 			mono_threads_pthread_kill (info, SIGTERM);
 
-			while (old_num_summarized == num_threads_summarized) {
+			// Number of seconds to give each dumping thread
+			int count = 20;
+
+			while (old_num_summarized == num_threads_summarized && count > 0) {
 				sleep (1);
 				mono_memory_barrier ();
-				/*mono_threads_pthread_kill (info, SIGTERM);*/
-
-				// Pause this handler so other handlers can run
-				/*int signum;*/
-				MOSTLY_ASYNC_SAFE_PRINTF("Waiting for signalled thread to collect stacktrace\n");
-				/*sigsuspend (&sigset);*/
-				/*mono_threads_pthread_kill (info, SIGTERM);*/
-				/*sigprocmask (SIG_UNBLOCK, &old_sigset, NULL);*/
-				/*g_assert (success == 0);*/
+				const char *name = thread_summary_state_to_str (summarizing_thread_state);
+				MOSTLY_ASYNC_SAFE_PRINTF("Waiting for signalled thread %zu to collect stacktrace. Status: %s\n", tid, name);
+				count--;
+			}
+			if (count == 0) {
+				MonoBoolean timeout_abort_success;
+
+				// After timeout
+				// Thread may have been in lock or something, may have died
+				// Either way, didn't respond. Rather than failing, we have to
+				// skip it.
+				mono_memory_barrier ();
+				switch (summarizing_thread_state) {
+				case MONO_SUMMARY_MUTATE_SHARED:
+					g_error ("Timed out when writing json log!");
+					break;
+
+				case MONO_SUMMARY_EMPTY: {
+					g_assert (summarizing_thread == (MonoNativeThreadId) NULL);
+					break;
+
+				case MONO_SUMMARY_EXAMINE:
+					MOSTLY_ASYNC_SAFE_PRINTF("Timed out, thread did not finish dumping\n");
+
+					MonoNativeThreadId old_val = ves_icall_System_Threading_Interlocked_CompareExchange_IntPtr ((gpointer) &summarizing_thread, GINT_TO_POINTER(tid) /* set */, GINT_TO_POINTER(NULL) /* compare */);
```

Comment:
Use `mono_atomic_cas_ptr`. `NULL` is already a pointer, no need to use `GINT_TO_POINTER` on it.